Add snap cards on bottom of the map with details about displayed hotels

jaroslav-kubicek opened the pull request:
![screen shot 2017-12-18 at 14 18 41](https://user-images.githubusercontent.com/2174629/34107950-6b11b692-e3fe-11e7-8f0e-245324f9f352.png)

*Note*  - there is one todo related to flow types to be done

Any feedback from user testing on how it feels when you are using it, is welcomed 🙂

mrtnzlml commented:
Please test these changes in [Expo](https://docs.expo.io/versions/latest/introduction/installation.html#mobile-client-expo-for-ios-and-android) after build has finished:

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=exp://exp.host/@mrtnzlml/code-review-map-swipe-view)

jaroslav-kubicek commented:
omg, FYI device orientation needs to be fixed...

mrtnzlml commented:
I don't know man. It feels so *not smooth*. I always drag like I would normally do but it waits for a while and that it finishes the movement with higher speed. Feels so weird.

Also, at some point, it scrolls to the very beginning (very fast). But if you continue long enough with scrolling it will never happen again (not nice).

Is the address not changing at all or is it still the same so I cannot see the effect?

> Device orientation needs to be fixed

Please do so.

mrtnzlml commented on app/common/src/map/Address.js:
```diff
@@ -0,0 +1,40 @@
+        </View>
+        <View>
+          <Text style={{ fontWeight: 'bold' }}>Address</Text>
+          <Text>950 Candice Wells Suite 375, Dublin</Text>
```
I would expect props here instead of text.

jaroslav-kubicek replied:
fixed

mrtnzlml commented on app/common/src/map/HotelSnap.js:
```diff
@@ -0,0 +1,58 @@
+  },
+});
+
+const HotelSnap = ({ width, hotel }: Props) => (
```
Can we find a better name for this? Other than "snap".

jaroslav-kubicek replied:
fixed (hopefully)

mrtnzlml commented on app/hotels/src/map/HotelSnaps.js:
```diff
@@ -0,0 +1,90 @@
+  };
+
+  onDimensionsChanged = ({ screen: { width } }: OnDimensionsChange) => {
+    this.setState({ availableWidth: width });
```
Why is this needed? Cannot we just use "width" from `Dimensions`? The trick is that you have to use it inside of render function so it gets updated every time you rerender (this is what you want anyway).

jaroslav-kubicek replied:
IMHO component is not re-rendered because the device has rotated, I even tried that, that's why I added lisneter, but I may miss something you already know...

mrtnzlml replied:
I did it here: https://github.com/kiwicom/react-native-app/blob/master/app/hotels/src/gallery/PhotosStripe.js#L23 (just to fix exactly the same issue)

mrtnzlml replied:
It's described here: https://facebook.github.io/react-native/docs/dimensions.html#get But now I really don't know why does it work... ❓

mrtnzlml replied:
Ok I bet it's about `onLayout` change in the `Swiper` so it doesn't work exactly how I thought... 🙂

mrtnzlml commented on app/hotels/src/map/HotelSnaps.js:
```diff
@@ -0,0 +1,90 @@
+  render = () => {
+    const { items, selectedIndex, onSnapToItem } = this.props;
+    const { availableWidth } = this.state;
+    const itemWidth = availableWidth * SNAP_WIDTH;
```
Seems like this is calculated twice (render and renderItem). Try to be DRY... 🙂

jaroslav-kubicek replied:
fixed

mrtnzlml commented on app/common/src/map/HotelCard.js:
```diff
@@ -0,0 +1,58 @@
+        <Image
+          style={styles.image}
+          source={{
+            uri: 'http://q.bstatic.com/images/hotel/square128/117/11753593.jpg',
```
Should I care now or we will connect the real data later?

jaroslav-kubicek replied:
I plan to do new PR for real data :)

jaroslav-kubicek commented:
> I don't know man. It doesn't feel very smooth. I always drag like I would normally do but it waits for a while and that it finishes the movement with higher speed. Feels so weird.

I removed momentum prop and added `decelerationRate`, which seems to be slightly smoother

> Also, at some point, it scrolls to the very beginning

valid point when `firstItem !== 0`, I expect `firstItem === 0` by default when you simply navigate to map by hitting "map" button in topbar, will investigate to make it work always.

Add resizing - I tried and it doesn't work without listeners :/

jaroslav-kubicek commented:
Related issue: https://github.com/archriss/react-native-snap-carousel/issues/232

mrtnzlml commented:
Let me know about the progress. It's not OK to merge obviously broken code...

jaroslav-kubicek commented:
workarounded :) cc @mrtnzlml

mrtnzlml commented on app/common/index.js:
```diff
@@ -20,16 +20,26 @@ export { default as NetworkImage } from './src/image/NetworkImage';
 
 export { default as PriceMarker } from './src/map/PriceMarker';
+export { default as HotelCard } from './src/map/HotelCard';
+export { default as Address } from './src/map/Address';
```
I do not think this should be in this shared package "just in case". I do not think there is high chance that we'll reuse these components. It's absolutely fine to create local components and use this only for true "shared" things. I am actually considering rename `common` to `shared` (like in React). I'll leave it here for now. I just want to mention it... 🙂

jaroslav-kubicek replied:
good point, I'm actually going to change it today :)

mrtnzlml commented:
> workarounded

I can reproduce it again just with a different flow. Try to click few times on different map markers and then try to swipe. It's broken again. I'll merge it for now because everyone who is not actively creating this project is bitching around but we will need to spend few more hours on this swiper. It's not very good yet.

mrtnzlml commented:
There is also this warning:

> VirtualizedList: You have a large list that is slow to update - make sure your renderItem function renders components that follow React performance best practices like PureComponent, shouldComponentUpdate, etc.
Object {
   "contentLength": 6003,
   "dt": 2661,
   "prevDt": 1851,
 }

mrtnzlml commented:
Merged.
